### PR TITLE
Implement UnitConverter class

### LINKED ISSUE #7: Create unit conversion class
### Purpose

Unit conversions will be a common operation across all nodes in the boat simulator. 

### Description

**Location:** `boat_simulator/common/unit_conversion.py`.
Create a class called `UnitConverter` that is able to convert a value when given the starting unit and end unit. Unit conversions should only be done on comparable units. For example, one would never convert meters to pounds -- it just doesn't make sense.

You may choose to store conversion rates in `boat_simulator/common/constants.py` or directly in the class. Whatever you believe is better for the implementation.

Make sure you test the class and make a pull request when it is ready.

#### Unit Types to Add
- Length (m, km, cm, etc.)
- Time (s, min, hr, etc.)
- Speed (m/s, kmph, miph, etc.)
- Acceleration (m/s^2, etc.)
- Angles (degrees and radians)

### PR DESCRIPTION
**Pull request by chrischang5**
### Description

- Resolves #7 

- Created `UnitConverter` class, `__init__` and `convert` functions
- Created `ScalarOrArray` type

### Verification

- Added tests in `tests/unit/common/test_unit_conversions.py`

### Resources

- https://www.calculatorsoup.com/calculators/conversions/acceleration.php
- https://www.unitconverters.net/weight-and-mass-converter.html
- Google Unit Converter

### CODE REVIEW

**Comment by chrischang5** [comment]
I really spelled `UnitConvertor` in my commit messages LOL my bad.

**Comment by chrischang5** [comment]
One step that could potentially be done to make a more complete test coverage is to parameterize tests for all unit conversions. Currently, only the angle to radians and radians to angle tests are parameterized in this way.

**Review by DFriend01 — CHANGES_REQUESTED** [review]
Looks great! There was one more thing that I wanted to check, and that was to see if the unit converter works with array-like objects as well as scalars.

Looking at the code, it probably should. Could you do the following:
- Add some tests to see if passing in ND arrays works where the conversion happens elementwise
- Change the typing from `Scalar` to `ArrayLike` for both the `UnitConverter` and the `ConversionFactor` class as appropriate? Take a look at the [ArrayLike type](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike). It covers both scalars and arrays in a single type.

File: tests/unit/common/test_unit_conversions.py
Comment (chrischang5) [inline tests/unit/common/test_unit_conversions.py:347]: Testing revealed regular Python lists cause unexpected results during conversion. Swapping `Scalar` to `numpy.typing.ArrayLike` was a plug-and-play process in terms of supporting `numpy` arrays. More work will have to be done to support Python lists alongside `numpy` arrays.

If no support for Python lists is not acceptable, then changes can be made but as of now, I have no knowledge of how to do this.

Comment (chrischang5) [reply tests/unit/common/test_unit_conversions.py:347]: More specifically, the lengths of lists seem to be altered during the unit conversion process (alongside getting potentially wrong results)

Comment (DFriend01) [reply tests/unit/common/test_unit_conversions.py:347]: I think we can assume that we would only care about using numpy arrays rather than lists or tuples too. If you can come up with a better way to type other than using `ArrayLike`, then I would be all ears.

Or even consider not typing at all until we get the mypy linter integrated into sailbot workspace because right now, we have no way of verifying that the type hints are valid all the time.

Comment (chrischang5) [reply tests/unit/common/test_unit_conversions.py:347]: Given the assumption that we are only using `numpy arrays`, what I have now could be sufficient. I think it could be deployed, but re-requesting review in case more thoughts come up.

Comment (DFriend01) [reply tests/unit/common/test_unit_conversions.py:347]: Taking a look at the numpy documentation, the other alternatives are not as great. This is probably good enough.

File: boat_simulator/common/unit_conversions.py
Comment (DFriend01) [inline boat_simulator/common/unit_conversions.py]: Since we only seem to be using `ArrayLike` from this package, it's probably cleaner to just import ArrayLike and change all the `npt.ArrayLike` lines to `ArrayLike`

EDIT: I found this type: https://numpy.org/doc/stable/reference/typing.html#numpy.typing.NDArray. Should we use this instead of `ArrayLike`? I am finding that many people are confusing this with "any array type", which is not always what we want. I want to use numpy arrays as much as possible. The `NDArray` type seems to assume that scalars are inside, so I believe that `NDArray` is enough.

**Review by DFriend01 — APPROVED** [review]
Good work. Just one small comment but I'm approving now. Once you make the change, feel free to merge this PR.

File: boat_simulator/common/unit_conversions.py
Comment (chrischang5) [reply boat_simulator/common/unit_conversions.py]: The type specified to `NDArray` must be a `numpy` recognized type. I don't think we can plug and play our own defined `Scalar` here. I tried myself and look like it's not happy. A list of accepted datatypes is here: https://numpy.org/doc/stable/reference/arrays.scalars.html#built-in-scalar-types.

You can specify `np.object_` as a datatype into `NDArray` (`NDArray[np.object_]` and then feed in `Scalars`. But, this seems like an over-engineered solution when `ArrayLike` exists.

There is a way to set `numpy` array datatypes to support ints or floats only, but it doesn't work with our `Scalar` object. The way to do it is outlined in the answer to [this StackOverflow post ](https://stackoverflow.com/questions/70714087/how-to-typehint-numpy-array-of-floats-and-or-integers), but it introduces more code. For these reasons, I'm just going to stick with `ArrayLike`.

Comment (DFriend01) [reply boat_simulator/common/unit_conversions.py]: You can try `Union[Scalar, NDArray]` if you are worried about typing.

Comment (DFriend01) [reply boat_simulator/common/unit_conversions.py]: You could also create a new custom type inside the boat simulator's typing module and import it from there. I think we would use `Union[Scalar, NDArray]` often.

Comment (chrischang5) [reply boat_simulator/common/unit_conversions.py]: The solution we came to is to create a new `ScalarOrArray` type for the `Union[Scalar, NDArray]` type.

**Pull request merged**
